src/java/org/apache/cassandra/locator/ReplicaLayout.java

```diff
@@ -18,364 +18,296 @@
      */
-    public E pending()
-    {
-        return pending;
-    }
-
-    public int blockFor()
-    {
-        return pending == null
-                ? consistencyLevel.blockFor(keyspace)
-                : consistencyLevel.blockForWrite(keyspace, pending);
-    }
-
-    public Keyspace keyspace()
+    public E all()
     {
-        return keyspace;
+        return natural;
     }
 
-    public ConsistencyLevel consistencyLevel()
+    public String toString()
     {
-        return consistencyLevel;
+        return "ReplicaLayout [ natural: " + natural + " ]";
     }
 
-    abstract public L withSelected(E replicas);
-
-    abstract public L withConsistencyLevel(ConsistencyLevel cl);
-
-    public L forNaturalUncontacted()
+    public static class ForTokenRead extends ReplicaLayout<EndpointsForToken> implements ForToken
     {
-        E more;
-        if (consistencyLevel.isDatacenterLocal() && keyspace.getReplicationStrategy() instanceof NetworkTopologyStrategy)
+        public ForTokenRead(EndpointsForToken natural)
         {
-            IEndpointSnitch snitch = keyspace.getReplicationStrategy().snitch;
-            String localDC = DatabaseDescriptor.getLocalDataCenter();
-
-            more = natural.filter(replica -> !selected.contains(replica) &&
-                    snitch.getDatacenter(replica).equals(localDC));
-        } else
+            super(natural);
+        }
+        @Override
+        public Token token()
         {
-            more = natural.filter(replica -> !selected.contains(replica));
+            return natural().token();
         }
 
-        return withSelected(more);
+        public ReplicaLayout.ForTokenRead filter(Predicate<Replica> filter)
+        {
+            EndpointsForToken filtered = natural().filter(filter);
+            if (filtered == natural()) return this;
+            return new ReplicaLayout.ForTokenRead(filtered);
+        }
     }
 
-    public static class ForRange extends ReplicaLayout<EndpointsForRange, ForRange>
+    public static class ForRangeRead extends ReplicaLayout<EndpointsForRange> implements ForRange
     {
-        public final AbstractBounds<PartitionPosition> range;
+        final AbstractBounds<PartitionPosition> range;
 
-        @VisibleForTesting
-        public ForRange(Keyspace keyspace, ConsistencyLevel consistencyLevel, AbstractBounds<PartitionPosition> range, EndpointsForRange natural, EndpointsForRange selected)
+        public ForRangeRead(AbstractBounds<PartitionPosition> range, EndpointsForRange natural)
         {
-            // Range queries do not contact pending replicas
-            super(keyspace, consistencyLevel, natural, null, selected);
+            super(natural);
             this.range = range;
         }
 
         @Override
-        public ForRange withSelected(EndpointsForRange newSelected)
+        public AbstractBounds<PartitionPosition> range()
```

Comment:
This seems to get copied around but is unused?

Reply:
It wasn't unused, but it can be made so - and it is now.  But for consistency with ReplicaLayout.ForTokenRead, I think it's fairly harmless.

A wider question is whether we actually want a ReplicaLayout for reads at all.  They're only a single Endpoints, and I do wonder if it's worth it.  But, again, for consistency with ReplicaPlan I think it's a very small price to pay.